### PR TITLE
Make sure fixedin field is lower than 256 chars

### LINKED ISSUE #1430: Bodhi sometimes puts too much text in the Bugzilla Fixed in Version field
bodhi-backend01 emitted an error e-mail yesterday due to putting too much text into the Fixed in Version field in Bugzilla:

```
Message
-------
[2017-04-10 21:20:21][     bodhi   ERROR]
python
Unable to close bug #1246113
```

Process Details
---------------
- host:     bodhi-backend01.phx2.fedoraproject.org
- PID:      64964
- name:     fedmsg-hub
- command:  /usr/bin/python /usr/bin/fedmsg-hub
- msg_id:   

Callstack that lead to the logging statement
--------------------------------------------
```
python
  File "/usr/lib64/python2.7/threading.py", line 777 in __bootstrap
    self.__bootstrap_inner()
  File "/usr/lib64/python2.7/threading.py", line 804 in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/site-packages/bodhi/server/consumers/masher.py", line 260 in run
    self.work()
  File "/usr/lib/python2.7/site-packages/bodhi/server/consumers/masher.py", line 351 in work
    self.modify_bugs()
  File "/usr/lib/python2.7/site-packages/bodhi/server/consumers/masher.py", line 57 in wrapper
    retval = method(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/bodhi/server/consumers/masher.py", line 811 in modify_bugs
    update.modify_bugs()
  File "/usr/lib/python2.7/site-packages/bodhi/server/models.py", line 1373 in modify_bugs
    bug.close_bug(self)
  File "/usr/lib/python2.7/site-packages/bodhi/server/models.py", line 2232 in close_bug
    bugs.bugtracker.close(self.bug_id, versions=versions, comment=self.default_message(update))
  File "/usr/lib/python2.7/site-packages/bodhi/server/bugs.py", line 145 in close
    log.exception("Unable to close bug #%d" % bug_id)

Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/bodhi/server/bugs.py", line 143, in close
    bug.close('ERRATA', **args)
  File "/usr/lib/python2.7/site-packages/bugzilla/bug.py", line 219, in close
    return self.bugzilla.update_bugs(self.bug_id, vals)
  File "/usr/lib/python2.7/site-packages/bugzilla/base.py", line 1264, in update_bugs
    return self._proxy.Bug.update(tmp)
  File "/usr/lib64/python2.7/xmlrpclib.py", line 1243, in __call__
    return self.__send(self.__name, args)
  File "/usr/lib/python2.7/site-packages/bugzilla/base.py", line 168, in _ServerProxy__request
    ret = ServerProxy._ServerProxy__request(self, methodname, params)
  File "/usr/lib64/python2.7/xmlrpclib.py", line 1602, in __request
    verbose=self.__verbose
  File "/usr/lib/python2.7/site-packages/bugzilla/base.py", line 259, in request
    return self._request_helper(url, request_body)
  File "/usr/lib/python2.7/site-packages/bugzilla/base.py", line 245, in _request_helper
    raise sys.exc_info()[1]
Fault: <Fault 104: 'The text you entered in the Fixed In Version field is too long (259 characters, above the maximum length allowed of 255 characters).'>
```

We should check for this max length, and probably should just leave off some NVRs if we are exceeding the length.

### PR DESCRIPTION
**Pull request by mattiaverga**
Make sure fixedin field is lower than 256 chars (fixes #1430 ).

Instead of adding the new version to the fixedin list and then check the length of the resultant string, check if the new version can be added to the string without going over the 255 chars limit.

### CODE REVIEW

**Comment by bowlofeggs** [comment]
jenkies test

File: bodhi/server/bugs.py [hunk 1]
```diff
@@ -201,15 +201,17 @@ def close(self, bug_id, versions, comment):
...
                 # structured.  We should use:
                 # - the full NVR as it appears in koji
                 # - space-separated if there's more than one.
-                args['fixedin'] = " ".join(fixedin)
+                fixedinString = " ".join(fixedin)
+                
+                # Add our build if its not already there
+                # but only if resultant string length is lower than 256 chars
+                if (version not in fixedin) and (len(fixedinString)+len(version) < 255):
+                    args['fixedin'] = fixedinString + " " + version
```
Comment (bowlofeggs) [inline bodhi/server/bugs.py]: This will prepend a space if ```fixedinString``` is the empty string, which will happen on the first item that adds a fixedin to a bug. I suggest something like this instead:

```
args['fixedin'] = ' '.join([fixedin_str, version])
```

File: bodhi/server/bugs.py [hunk 2]
```diff
@@ -201,15 +201,17 @@ def close(self, bug_id, versions, comment):
...
-                # And add our build if its not already there
-                if version not in fixedin:
-                    fixedin.append(version)
 
                 # There are Red Hat preferences to how this field should be
                 # structured.  We should use:
                 # - the full NVR as it appears in koji
                 # - space-separated if there's more than one.
-                args['fixedin'] = " ".join(fixedin)
+                fixedinString = " ".join(fixedin)
```
Comment (bowlofeggs) [inline bodhi/server/bugs.py]: For [PEP-8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names), this should be named ```fixedin_str``` (or ```fixedin_string``` if you prefer the full name).

File: bodhi/server/bugs.py [hunk 3]
```diff
@@ -201,15 +201,17 @@ def close(self, bug_id, versions, comment):
...
+                
+                # Add our build if its not already there
+                # but only if resultant string length is lower than 256 chars
+                if (version not in fixedin) and (len(fixedinString)+len(version) < 255):
```
Comment (bowlofeggs) [inline bodhi/server/bugs.py]: Also for [PEP-8](https://www.python.org/dev/peps/pep-0008/#other-recommendations), there should be spaces surrounding the ```+``` operator.

File: bodhi/server/bugs.py [hunk 4]
```diff
@@ -201,15 +201,17 @@ def close(self, bug_id, versions, comment):
...
 
                 # There are Red Hat preferences to how this field should be
                 # structured.  We should use:
                 # - the full NVR as it appears in koji
                 # - space-separated if there's more than one.
-                args['fixedin'] = " ".join(fixedin)
+                fixedinString = " ".join(fixedin)
+                
+                # Add our build if its not already there
+                # but only if resultant string length is lower than 256 chars
```
Comment (bowlofeggs) [inline bodhi/server/bugs.py]: It might be good to put a link to the bug report in this comment block so future readers know where the 256 limit comes from.

**Review by bowlofeggs — CHANGES_REQUESTED** [review]
Hello, and thanks for the patch! I've left a few notes of changes we'll want to make. In addition to those changes:

* We will want to add tests for this change to ```bodhi/tests/server/test_bugs.py```. I suggest a test that makes sure we don't exceed the 255 limit, and also a test to ensure we don't prepend an empty space on the first fixedin version.
* Please squash your two commits into one commit.
* Please move the ```fixes #1430``` onto its own line in your commit message. This way GitHub will automatically close the bug when this gets merged, which is handy ☺
* Please sign the [DCO](https://developercertificate.org/) in your commit message. The way to do that is described in Bodhi's [contribution guidelines](https://bodhi.fedoraproject.org/docs/developer/index.html#contribution-guidelines).

File: bodhi/server/bugs.py [hunk 5]
```diff
@@ -201,15 +201,17 @@ def close(self, bug_id, versions, comment):
...
-                if version not in fixedin:
-                    fixedin.append(version)
 
                 # There are Red Hat preferences to how this field should be
                 # structured.  We should use:
                 # - the full NVR as it appears in koji
                 # - space-separated if there's more than one.
-                args['fixedin'] = " ".join(fixedin)
+                fixedinString = " ".join(fixedin)
+
```
Comment (bowlofeggs) [inline bodhi/server/bugs.py]: ```flake8``` fails on this line because it contains whitespace:

```
rawhide	./bodhi/server/bugs.py:210:1: W293 blank line contains whitespace
rawhide	                
rawhide	^
```

If you are using the Vagrant development environment, you can reproduce this test by running ```btest```. You can also run ```flake8``` directly if you like.

File: bodhi/server/bugs.py (hunk 3 above)
Comment (bowlofeggs) [reply bodhi/server/bugs.py]: ```flake8``` mentioned this too in the CI test run:

```
rawhide	./bodhi/server/bugs.py:213:68: E226 missing whitespace around arithmetic operator
rawhide	                if (version not in fixedin) and (len(fixedinString)+len(version) < 255):
rawhide	     
```

**Comment by mattiaverga** [comment]
I'm having some problems creating the tests: I have to figure out how to access the "fixedin" field of the bug.
I tried to test with 
`self.assertEqual(bz._bz.getbug.return_value.fixedin, 'bodhi-3.1.0-1.fc27')`
but it seems that it's not so easy...

**Comment by mattiaverga** [comment]
While adding new tests I found that the close() method was only tested for failing, not for success.
So, I wrote a simple test case starting from the existent one, tested against develop and I found that is failing:
```
    @mock.patch('bodhi.server.bugs.log.exception')
    def test_close_successful(self, exception):
        """Test the close() method with a success case."""
        bz = bugs.Bugzilla()
        bz._bz = mock.MagicMock()
        bz._bz.getbug.return_value.component = 'bodhi'

        bz.close(12345, {'bodhi': ['bodhi-3.1.0-1.fc27']}, 'Fixed. Closing bug.')

        bz._bz.getbug.assert_called_once_with(12345)
        bz._bz.getbug.return_value.close.assert_called_once_with('ERRATA',
                                                                comment='Fixed. Closing bug.',
                                                                fixedin='bodhi-3.1.0-1.fc27')
        self.assertEqual(exception.call_count, 0)
```
It fails with:
```
  File "bodhi/server/bugs.py", line 215, in close
    args['fixedin'] = " ".join(fixedin)
TypeError: sequence item 0: expected string, list found
```
In fact, the problem is in the call `bz.close(12345, {'bodhi': ['bodhi-3.1.0-1.fc27']}, 'Fixed. Closing bug.')` that should be `bz.close(12345, {'bodhi': 'bodhi-3.1.0-1.fc27'}, 'Fixed. Closing bug.')`.

So, my question is: is the call in the existent test wrong? Or the versions dict element is supposed to be a list?

**Comment by mattiaverga** [comment]
I'm starting to getting lost here...
Looking at models.py, the `versions` object passed to the close() function is:
`versions = {'component' : ['version', ]}`
so in bodhi/server/bugs.py at line 199 the code
`version = versions[bug.component]`
means we have `version = ['version', ]`

After that we do:
```
if version not in fixedin:
    fixedin.append(version)
...
args['fixedin'] = " ".join(fixedin)
```
I can't see how this could work, since we are joining strings with a list of strings. In fact running this code in a console fails just like the testcase I wrote.
BUT it seems that on the real istance of Bodhi this is working without errors... looking at recently closed bugs I can see the fixedin field is populated.

**Comment by mattiaverga** [comment]
My fault, the `versions` object is:
versions = {'component' : 'version', }
so it was the call in `test_close_fault` that was wrong, but obviously it was passing because it was triggering an Exception...

Now all the tests pass.

**Comment by bowlofeggs** [comment]
jenkies test

**Comment by bowlofeggs** [comment]
Hi @mattiaverga! Sorry I've not been very responsive. The past couple of weeks have been really crazy due to the F28 Bodhi activation and I've been working long hours on problems related to that. It sounds like you were able to get some of your questions answered though!

I think CI might be broken right now (also due to some F28 stuff) so Jenkins might say this fails, but I'm running it to see if it's fixed today.

**Comment by bowlofeggs** [comment]
I can confirm that the tests pass locally on non-rawhide containers.

**Review by bowlofeggs — APPROVED** [review]
Looks good, thanks for the patch!

I think rawhide should be fixed tomorrow, so I will wait to merge until CI can formally pass, but I think it'll be fine.

**Comment by bowlofeggs** [comment]
I opened a PR to create one more test related to this patch: https://github.com/fedora-infra/bodhi/pull/2205

I'll rebase and merge it after this one is merged (I confirmed that it passes with this code).

**Comment by bowlofeggs** [comment]
Let's see if Rawhide is working yet :)

**Comment by bowlofeggs** [comment]
jenkies test

**Comment by bowlofeggs** [comment]
well lookee there - rawhide works again. Thanks for the patch!
